File: LeetCode/C/98-tree-验证二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */
#include "main.h"
// @lc code=start
class Solution {
  public:
    bool isValidBST(TreeNode *root) { return check(root, LONG_MIN, LONG_MAX); }

    bool check(TreeNode *root, long low, long high) {
        if (root == nullptr)
            return true;
        if (root->val <= low || root->val >= high)
            return false;
        return check(root->left, low, root->val) &&
               check(root->right, root->val, high);
    }
};
// @lc code=end

File: LeetCode/Go/98-tree-验证二叉搜索树.go
/*
 * @lc app=leetcode.cn id=98 lang=golang
 *
 * [98] 验证二叉搜索树
 */

// package leetcode

// @lc code=start
import "math"

func isValidBST(root *TreeNode) bool {
	var check func(*TreeNode, int, int) bool
	check = func(node *TreeNode, low int, high int) bool {
		if root == nil {
			return true
		}
		if root.Val <= low || root.Val >= high {
			return false
		}
		return check(root.Left, low, root.Val) && check(root.Right, root.Val, high)
	}

	return check(root, math.MinInt64, math.MaxInt64)
}

// @lc code=end
